### PR TITLE
Creating an S3 bucket to store waf logs

### LINKED ISSUE #120: waf logs s3 bucket
## Type of Issue:

- [ ] Bug: Something isn't working
- [x] Enhancement: New feature or request
- [ ] Question: Further information is requested (I have a question)

#### Issue Creator Checklist

- [x] The appropriate label (bug, enhancement, question) has been attached to this issue.
- [x] This issue has been thoroughly documented below; a developer should be able to understand the issue by reading it.

## Background

The purpose of this ticket is to provide an s3 bucket to write WAF logs. The reason for this is in the Greenfiled accounts that are SPLUNK integrated the bucket is created for us by an outside team. The purpose of this is to programmatically create the bucket in a non SPLUNK integrated environment.  This gives us an s3 bucket to write WAF logs to in the quickstart and reference in later services that actually write the logs.

## AC:

- S3 bucket will be created as a serverless service during deployment  to hold waf logs.

### PR DESCRIPTION
**Pull request by joeaero19**
## Purpose

We want to have WAF logging enabled for all projects to get their logs to Splunk. This bucket will hold all the waf logs generated. From here the logs will be shipped off to splunk.
This PR only creates the  s3 bucket to store waf logs. 

#### Linked Issues to Close

Closes #120 

## Approach

For some CMS applications, the logs are being written to an existing s3 bucket created by the CMS infra team. If so, we will be injecting the bucket name via SSM.
In other cases like the quickstart, We will need to create and export the the bucket info (ARN) to cloudformation output. By making the s3 bucket its own serverless service, there is a clean and smooth interaction between ui and the bucket.

#### Pull Request Creator Checklist

- [x] This PR has an associated issue or issues.
- [x] The associated issue(s) are linked above.
- [x] This PR meets all acceptance criteria for those issues.
- [x] This PR and linked issue(s) are adequately documented
- [x] This PR and linked issues(s) are a complete description of the changeset; an individual or team should be able to understand the issue(s) and changes by reading through this PR and it's links, with no further interaction.
- [x] Someone has been assigned this PR.
- [x] At least one person has been marked as reviewer on this PR.

#### Pull Request Reviewer/Assignee Checklist

- [x] This PR has an associated issue or issues.
- [x] The associated issue(s) are linked above.
- [x] This PR meets all acceptance criteria for those issues.
- [x] This PR and linked issue(s) are adequately documented
- [x] This PR and linked issues(s) are a complete description of the changeset; an individual or team should be able to understand the issue(s) and changes by reading through this PR and it's links, with no further interaction.

### CODE REVIEW

File: services/ui/serverless.yml [hunk 1]
```diff
@@ -155,6 +155,7 @@ resources:
           HostedZoneId: Z2FDTNDATAQYW2
           EvaluateTargetHealth: false
         Type: A
+
```
Comment (mdial89f) [inline services/ui/serverless.yml:158]: Undo this new line

File: services/ui-waflog-s3-bucket/.serverless/cloudformation-template-create-stack.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/.serverless/cloudformation-template-create-stack.json:1]: The .serverless directory should be git ignored.  Any commited files in .serverless should be removed.

File: services/ui-waflog-s3-bucket/.serverless/cloudformation-template-update-stack.json [hunk 3]
```diff
@@ -0,0 +1,93 @@
+{
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/.serverless/cloudformation-template-update-stack.json:1]: The .serverless directory should be git ignored.  Any commited files in .serverless should be removed.

File: services/ui-waflog-s3-bucket/.serverless/serverless-state.json [hunk 4]
```diff
@@ -0,0 +1,243 @@
+{
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/.serverless/serverless-state.json:1]: The .serverless directory should be git ignored.  Any commited files in .serverless should be removed.

File: services/ui-waflog-s3-bucket/package.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/package.json:1]: There should be a yarn.lock file checked in.  It's missing.

**Review by mdial89f — CHANGES_REQUESTED** [review]

**Comment by mdial89f** [comment]
The Purpose section needs to have more detail around what's trying to be accomplished and why.
The Approach section should include any implementation details or hurdles.  In particular, explaining why this bucket is in a separate service is important... without an explanation, it's odd that it's in its own service.
Remove the instructions from the PR (text you see highlighted... when editing, it's text that begins and ends with underscores).

**Comment by mdial89f** [comment]
The title could be more descriptive.

File: services/ui-waflog-s3-bucket/.serverless/cloudformation-template-create-stack.json (hunk 2 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/.serverless/cloudformation-template-create-stack.json:1]: done

File: services/ui-waflog-s3-bucket/.serverless/cloudformation-template-update-stack.json (hunk 3 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/.serverless/cloudformation-template-update-stack.json:1]: done

File: services/ui-waflog-s3-bucket/.serverless/serverless-state.json (hunk 4 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/.serverless/serverless-state.json:1]: done

File: services/ui-waflog-s3-bucket/package.json (hunk 5 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/package.json:1]: added file

File: services/ui/serverless.yml (hunk 1 above)
Comment (joeaero19) [reply services/ui/serverless.yml:158]: done

**Review by dwhitecl — APPROVED** [review]

File: services/ui-waflog-s3-bucket/serverless.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+provider:
+  name: aws
+  runtime: nodejs12.x
+  region: us-east-1
+
+custom:
+  stage: ${opt:stage, self:provider.stage}
+  serverlessTerminationProtection:
+    stages:
+      - master
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/serverless.yml:17]: @joeaero19 Add the standard list of protected stages.  Any of the other services will have the full list in their serverless.yml
```    stages: # This is a list of common names for important envs that should not be destroyed.  You can remove the stage names your project doesn't use; this list is meant to be inclusive.
      - master
      - val
      - production
      - develop
      - main
      - impl
      - val
      - prod``

File: services/ui-waflog-s3-bucket/serverless.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      - val
+      - production
+
+resources:
+  Resources:
+    WaflogsUploadBucket:
+      Type: AWS::S3::Bucket
+      Properties:
+        BucketName: ${self:custom.stage}-waflogsbucket
```
Comment (mdial89f) [inline services/ui-waflog-s3-bucket/serverless.yml:26]: This is problematic, as S3 is a global name space.
I recommend using a random prefix, or the account id, at the start of the bucket name.
```BucketName:  ${AWS::AcountId}-${self:service.name}-${self:custom.stage}-waflogs```

**Review by mdial89f — CHANGES_REQUESTED** [review]

File: services/ui-waflog-s3-bucket/serverless.yml (hunk 6 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/serverless.yml:17]: done.

File: services/ui-waflog-s3-bucket/serverless.yml (hunk 7 above)
Comment (joeaero19) [reply services/ui-waflog-s3-bucket/serverless.yml:26]: done

**Review by mdial89f — APPROVED** [review]

**Pull request merged**